smslibrary/src/main/java/com/eis/smslibrary/SMSHandler.java

```diff
@@ -1,13 +1,149 @@
 package com.eis.smslibrary;
 
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
 import com.eis.communication.CommunicationHandler;
-import com.eis.communication.Message;
+import com.eis.smslibrary.listeners.SMSReceivedListener;
+import com.eis.smslibrary.listeners.SMSSentListener;
+
+import java.lang.ref.WeakReference;
+
+/**
+ * Communication handler for SMSs. It's a Singleton, you should
+ * access it with {@link #getInstance}, and before doing anything you
+ * should call {@link #setup}.
+ *
+ * @author Luca Crema, Marco Mariotto, Alberto Ursino, Marco Tommasini
+ */
+public class SMSHandler implements CommunicationHandler<SMSMessage> {
+
+    public static final String SENT_MESSAGE_INTENT_ACTION = "SMS_SENT";
+
+    /**
+     * Singleton instance
+     */
+    private static SMSHandler instance;
+
+    /**
+     * Weak reference doesn't prevent garbage collector to
+     * de-allocate this class when it has reference to a
+     * context that is still running. Prevents memory leaks.
+     */
+    private WeakReference<Context> context;
```

Comment:
I just learned this and it saves a whole lot of trouble. Big thanks

Reply:
Learned about it this morning too